test/test_cdd.cpp

```diff
@@ -1004,7 +1004,8 @@ TEST_CASE("CDD BDD to array test")
         int32_t expected_vars1[8] = {bdd_start_level, default_value,       default_value,       default_value,
                                      bdd_start_level, bdd_start_level + 1, bdd_start_level + 2, default_value};
         REQUIRE(equal(array1.vars, expected_vars1, 8));
-        int32_t expected_values1[8] = {1, default_value, default_value, default_value, 0, 1, 1, default_value};
+        int32_t expected_values1[8] = {1, default_value, default_value, default_value,
```

Comment:
If those were `std::vector` instead of `int32[]` then one could use `REQUIRE(expected_values == array1.values)`, then `equal(ptr1, ptr2, size)` would be redundant and `doctest` would be able to print the values in case the equality fails.

Reply:
Also taking your previous comment into account, I'm now thinking of the following concept. What if we would change the output of `cdd_bdd_to_array` to a C++ struct (i.e., using `std::vector`) and create a new function that calls `cdd_bdd_to_array` and convert the result to the current `bdd_arrays` C-struct?
Then we can also test the core functionality using the C++ data types and get rid of the `equals` function.

Reply:
You can add C++ API function surrounded by `#ifdef __cplusplus` and `#endif` and then use it in tests if you want.
Once the code gets `std::vector` then it becomes its owner, then one can call `.data()` to access raw pointer and use it in C context, but we cannot really wrap it and create a C function returning raw pointer unless we maintain that vector separately, because such a function would deallocate the memory once it returned the raw pointer :-(
Also `bdd_arrays` store a pointer to an array. It would be nice to store the `std::vector` instead, but then all the C code would break.

The whole memory management strategy needs to be rethought. IMHO most of the kernel code concerned with hash-tables and memory needs to go.